### DISCUSSION TITLE
Best way to check if sdcard exists

### DISCUSSION THREAD

**Discussion by brad28b** [apache/cordova#449] [Q&A]
I am using the 'File' plugin in my app, and it downloads some files to the device from a remote web server. I am trying to work out how to check if an sdcard exists in the device. 

If I test this on a emulator without an sdcard, I was hoping that the value of cordova.file.externalRootDirectory would be null, which would mean I could use that, but its not null, even without a sdcard in the device, cordova.file.externalRootDirectory has a value of 'file:///storage/emulated/0/'

How best to check for existence of sdcard?

**Comment by breautek** [comment]
Android abstracts the presence of the sdcard. There is a partition that acts as an external storage medium when the sdcard is not present, which is generated represented as `file:///storage/emulated/<userID>/` where the userID is generally `0` unless if you're on a corporate issued device where they manage users on devices.

I think if there is a physical sdcard present then the path will be `file:///sdcard` but that's an implementation detail and may not be true in all circumstances. The file plugin I think only does this check at startup, so removing the sdcard while the app is open, or inserting one in while the app is open probably won't change the `externalRootDirectory` path. You should be able to test this with an emulator as well in the emulator settings. By default the simulator is configured to act as if there is no attached sd card.

There is a lot of devices available that doesn't have any sdcard slot, so it's unwise to rely on the presence sdcards. These devices still emulates external storage however.

**Reply by brad28b** [reply]
> There is a lot of devices available that doesn't have any sdcard slot, so it's unwise to rely on the presence sdcards. These devices still emulates external storage however.

Yes that is what I am trying to do, check if sdcard exists, then if it does, check sdcard available free space, then check available free space on internal, and work out the best location to store the files. My preference is to store internal, most modern devices have large enough internal storage to not bother with sdcards. But, I have to account for a case where for example an scard is present, there is not enough space on internal, but there is on sdcard, then I would store data there.

**Reply by breautek** [reply]
> But, I have to account for a case where for example an scard is present, there is not enough space on internal, but there is on sdcard, then I would store data there.

Checking the external storage should be sufficient unless if you want to check specifically for and use a removable drive.

The file plugin if I recall correctly only concerns itself with the internal partition, and the primary external partition. That is, it doesn't know if the primary external partition is a sdcard or an emulated partition. In the case where there is a sdcard present, there will be multiple external volumes, but the file plugin only uses the primary one I'm pretty sure, which may or may not be the sdcard.

All phones today (including all devices that Cordova supports), afaik, will have an isolated data partition to emulate the external storage medium (which is the `/storage/emulated/0/` path).

Taken from the android [docs](https://developer.android.com/reference/android/os/Environment#getExternalStorageDirectory()):

> Note: don't be confused by the word "external" here. This directory can better be thought as media/shared storage. It is a filesystem that can hold a relatively large amount of data and that is shared across all applications (does not enforce permissions). Traditionally this is an SD card, but it may also be implemented as built-in storage in a device that is distinct from the protected internal storage and can be mounted as a filesystem on a computer.

If you want to check every external storage volume, then I don't think that's possible via the file plugin since it only uses the primary external volume.
